contracts/upgradeable_contracts/ERC677Bridge.sol

```diff
@@ -15,16 +16,17 @@ contract ERC677Bridge is BasicBridge {
     }
 
     function onTokenTransfer(address _from, uint256 _value, bytes /*_data*/) external returns(bool) {
-        require(msg.sender == address(erc677token()));
+        IBurnableMintableERC677Token token = erc677token();
```

Comment:
Since this contract is assumed to be common for other contracts covering the functionality to bridge erc677 tokens, it should work with the `ERC677` instance rather than `BurnableMintableERC677Token`.

Reply:
For method `erc677token()` wee need to return `IBurnableMintableERC677Token` because `HomeBridgeErcToErc` and `ForeignBridgeNativeToErc` calls `erc677token().mint()` which is not defined on `ERC677`

If we use `ERC677` on `onTokenTransfer` method, also `ERC677` should be used on `bridgeSpecificActionsOnTokenTransfer(ERC677 _token, address _from, uint256 _value)`

But then on `ERC677BridgeForBurnableMintableToken` contract, method `bridgeSpecificActionsOnTokenTransfer(IBurnableMintableERC677Token _token, address _from, uint256 _value)`
is not overriding the method defined on `ERC677Bridge` but interpreted as a different method which cause incorrect behaviour.

Perhaps an option could be to instead of returning/receiving `ERC677` or `IBurnableMintableERC677Token` we can return and receive parameters as `address` and then convert to the proper contract to call the needed methods

Reply:
Could you consider such kind of changes as a suggestion to address your comments?
```diff
diff --git a/contracts/upgradeable_contracts/ERC677Bridge.sol b/contracts/upgradeable_contracts/ERC677Bridge.sol
index 3d99026..37777b3 100644
--- a/contracts/upgradeable_contracts/ERC677Bridge.sol
+++ b/contracts/upgradeable_contracts/ERC677Bridge.sol
@@ -5,7 +5,6 @@ import "./BasicBridge.sol";
 import "../ERC677.sol";
-import "../IBurnableMintableERC677Token.sol";
 
 contract ERC677Bridge is BasicBridge {
-    function erc677token() public view returns(IBurnableMintableERC677Token) {
-        return IBurnableMintableERC677Token(addressStorage[keccak256(abi.encodePacked("erc677token"))]);
+    function erc677token() public view returns(ERC677) {
+        return ERC677(addressStorage[keccak256(abi.encodePacked("erc677token"))]);
     }
@@ -18,3 +17,3 @@ contract ERC677Bridge is BasicBridge {
     function onTokenTransfer(address _from, uint256 _value, bytes /*_data*/) external returns(bool) {
-        IBurnableMintableERC677Token token = erc677token();
+        ERC677 token = erc677token();
         require(msg.sender == address(token));
@@ -26,3 +25,3 @@ contract ERC677Bridge is BasicBridge {
 
-    function bridgeSpecificActionsOnTokenTransfer(IBurnableMintableERC677Token _token, address _from, uint256 _value) internal {
+    function bridgeSpecificActionsOnTokenTransfer(ERC677 _token, address _from, uint256 _value) internal {
         fireEventOnTokenTransfer(_from, _value);
diff --git a/contracts/upgradeable_contracts/ERC677BridgeForBurnableMintableToken.sol b/contracts/upgradeable_contracts/ERC677BridgeForBurnableMintableToken.sol
index f93e841..3483626 100644
--- a/contracts/upgradeable_contracts/ERC677BridgeForBurnableMintableToken.sol
+++ b/contracts/upgradeable_contracts/ERC677BridgeForBurnableMintableToken.sol
@@ -8,4 +8,4 @@ import "../IBurnableMintableERC677Token.sol";
 contract ERC677BridgeForBurnableMintableToken is ERC677Bridge {
-    function bridgeSpecificActionsOnTokenTransfer(IBurnableMintableERC677Token _token, address _from, uint256 _value) internal {
-        _token.burn(_value);
+    function bridgeSpecificActionsOnTokenTransfer(ERC677 _token, address _from, uint256 _value) internal {
+        IBurnableMintableERC677Token(_token).burn(_value);
         fireEventOnTokenTransfer(_from, _value);
diff --git a/contracts/upgradeable_contracts/erc20_to_erc20/HomeBridgeErcToErc.sol b/contracts/upgradeable_contracts/erc20_to_erc20/HomeBridgeErcToErc.sol
index 6fb9992..af2b3db 100644
--- a/contracts/upgradeable_contracts/erc20_to_erc20/HomeBridgeErcToErc.sol
+++ b/contracts/upgradeable_contracts/erc20_to_erc20/HomeBridgeErcToErc.sol
@@ -63,3 +63,3 @@ contract HomeBridgeErcToErc is ERC677Receiver, EternalStorage, BasicBridge, Basi
         setTotalExecutedPerDay(getCurrentDay(), totalExecutedPerDay(getCurrentDay()).add(_value));
-        return erc677token().mint(_recipient, _value);
+        return IBurnableMintableERC677Token(erc677token()).mint(_recipient, _value);
     }
diff --git a/contracts/upgradeable_contracts/native_to_erc20/ForeignBridgeNativeToErc.sol b/contracts/upgradeable_contracts/native_to_erc20/ForeignBridgeNativeToErc.sol
index 7bdd35e..70f1d0c 100644
--- a/contracts/upgradeable_contracts/native_to_erc20/ForeignBridgeNativeToErc.sol
+++ b/contracts/upgradeable_contracts/native_to_erc20/ForeignBridgeNativeToErc.sol
@@ -53,3 +53,3 @@ contract ForeignBridgeNativeToErc is ERC677Receiver, BasicBridge, BasicForeignBr
     function claimTokensFromErc677(address _token, address _to) external onlyIfOwnerOfProxy {
-        erc677token().claimTokens(_token, _to);
+        IBurnableMintableERC677Token(erc677token()).claimTokens(_token, _to);
     }
@@ -58,3 +58,3 @@ contract ForeignBridgeNativeToErc is ERC677Receiver, BasicBridge, BasicForeignBr
         setTotalExecutedPerDay(getCurrentDay(), totalExecutedPerDay(getCurrentDay()).add(_amount));
-        return erc677token().mint(_recipient, _amount);
+        return IBurnableMintableERC677Token(erc677token()).mint(_recipient, _amount);
     }
```
If they are OK I could commit them to the branch `175-ERC677-support-foreign-erc20-to-erc20` and you can double check after that that everything is ready for merge

Reply:
Yes, looks good to me! Thanks!

Reply:
I pushed the changes. Please double check and test them.

Reply:
Changes looks good and all tests are passing so I think we can merge this